### PR TITLE
Advanced pickit

### PR DESCRIPTION
**Pull request by RB-Development**
**To use the advanced itemfilter an ID-Tome needs to be in the inventory.**

This PR allows you to decide if magic, rare and uniques should be stashed or not.
Furthermore you are now able to specify nested filters in your pickit.ini
(Examples are in pickit.ini)

Added several item properties in the assets folder.

### CODE REVIEW

**Comment by aeon0** [comment]
GJ on the effort. I am having a few conceptional issues with it though. There is already an "id items at cain" feature. Imho this is a lot more "safe" and reliable. We all know the issues we can face when doing stuff in the inventory...
Also I am not a big fan of the yml file which adds yet another layer of pickit info to a different place in a different format. Why cant we add these values just like we do with the socket info or etheral info in the pickit.ini?

**Comment by RB-Development** [comment]
> 

Havent mentioned the "id @ Cain" feature. I will give it a try and implement the routine instead of the ID-Tome.
Just need to implement some logic while picking up items if an identification "@Cain" ist needed.
Ill try to make the attributes availible in the pickit.ini file so we dont need a separate file. Hope it wont get to messy :)

**Comment by RB-Development** [comment]
Integrated yaml informationen into pickit.ini. No yaml file needed.
Integrated routine to check if identification is activated else the include for magic, rare, sets and uniques are disabled.
Only donwside: no nested attributation in pickit.ini possible

File: config/pickit.ini
Comment (aeon0) [inline config/pickit.ini:181]: So we dont allow +3 trap, +3 lightning? Maybe only filter for +3_lightning?

File: config/pickit.ini
Comment (aeon0) [inline config/pickit.ini:211]: puh. There can be really good stats rings without fcr or ll, ml. Rings don't take up much space, I'd vote to just keep them all.

File: config/pickit.ini
Comment (aeon0) [inline config/pickit.ini:214]: Can rare circlets get 3_pcombat? I thought only magic can.

File: src/bot.py
Comment (aeon0) [inline src/bot.py:111]: Any reason to set this to 5?

**Review by aeon0 — CHANGES_REQUESTED** [review]
And obviously a bit of clean up, removing yml and commented code. But its going in the right direction. Good job.

File: config/pickit.ini
Comment (RB-Development) [reply config/pickit.ini]: changed to pickup 2 assa, 3 Lightning and traps+ lightning sentry.
((2_assasin_skills, 3_lightning_sentry), (traps, lightning_sentry))

File: config/pickit.ini
Comment (RB-Development) [reply config/pickit.ini]: Removed that include filter in this case

File: config/pickit.ini
Comment (RB-Development) [reply config/pickit.ini]: The cant. Removed that part

File: src/bot.py
Comment (RB-Development) [reply src/bot.py]: Just for Debug to trigger stashing when bot start. Set to 0 again

**Review by aeon0 — CHANGES_REQUESTED** [review]
I will wait with merging for next release because I would see this PR as super nice, but also somewhat risky to have unexpected behaviours. So the course of actions would be, do a new release, merge this and test it properly on master and get ready for the next one.

File: src/ui/ui_manager.py
Comment (RB-Development) [reply src/ui/ui_manager.py]: Y, working fine. tested :)

**Comment by RB-Development** [comment]
> I will wait with merging for next release because I would see this PR as super nice, but also somewhat risky to have unexpected behaviours. So the course of actions would be, do a new release, merge this and test it properly on master and get ready for the next one.

Totally agree. Im also working on a gambling option. Right now an alpha version is running but needs some small improvements before i can start another PR.

File: config/pickit.ini
Comment (aeon0) [inline config/pickit.ini:45]: If you make misc_gold default 1 I would also make gold_trav_only param in [car] section default 1

**Review by aeon0 — CHANGES_REQUESTED** [review]
I wouldn't give the code any elegance award, kind of brute force, but I'd be ok with it given the fact that we plan to rewrite most of the item pickit with ocr and a more detailed item description format.
But one thing I would ask for are unittests for the parse function. It seems like code that could 1) be easily unittested 2) would be very beneficial for it.

**Comment by RB-Development** [comment]
Some small changes made to make unittests comftable. Unittests added, small adjustment regarding pickit

**Comment by aeon0** [comment]
@RB-Development In an attempt to solve the merge conflict on the github GUI I think I broke sth maybe? Can you check? Otherwise I'd fine merging.

**Comment by RB-Development** [comment]
Looks like you killed the function in config.py to read the nested filters
`def parse_item_config_string(self, key: str = None) -> ItemProps:
        string = self._select_val("items", key).upper()
        return self.string_to_item_prop (string)

    def string_to_item_prop (self, string: str) -> ItemProps:
        item_props = ItemProps()
        brk_on = 0
        brk_off = 0
        section = 0
        counter = 0
        start_section = 0
        start_item = 0
        include_list = []
        exclude_list = []
        for char in string:
            new_section = False
            counter+=1
            if char == "(":
                brk_on +=1
            elif char == ")":
                brk_off += 1
            if ((char == "," and brk_on==brk_off)):
                new_section = True
                if  (counter == len (string)):
                    string_section = string [start_section:counter]
                else:
                    string_section = string [start_section:counter-1]
                if section == 0:
                    item_props.pickit_type = int (string_section)
                section +=1
                start_section = counter
            if ((char == "," and (brk_on==brk_off+1)) or new_section or counter == len (string)):
                if new_section: 
                    section -=1
                if start_item ==0:
                    start_item = start_section
                if  (counter == len (string)):
                    item = string [start_item:counter]
                else:
                    item = string [start_item:counter-1]
                if section == 0 and counter == len (string):
                    item_props.pickit_type = int (item)
                    pass
                if section ==1:
                    include_list.append (item)
                    start_item = counter +1
                elif section ==2:
                    exclude_list.append (item)
                    start_item = counter +1
                if new_section: 
                    section +=1
        if (len (include_list)>0 and (len (include_list[0]) >=6)):
            if ("AND" in include_list[0][0: 6] and not ")" in include_list[0]):
                item_props.include_type = "AND"
            else:
                item_props.include_type = "OR"
        if (len (exclude_list)>0 and (len (exclude_list[0]) >=6)):
            if ("AND" in exclude_list[0][0:6] and not ")" in include_list[0]):
                item_props.exclude_type = "AND"
            else:
                item_props.exclude_type = "OR"
        for i in range (len(include_list)):
            include_list[i]  = include_list[i].replace (" ","").replace ("OR(","").replace ("AND(", "").replace ("(", "").replace (")","")
            include_list[i]  = include_list[i].split (",")
        for l in range (len (exclude_list)):
            exclude_list[l] = exclude_list[l].replace (" ","").replace ("OR(","").replace ("AND(", "").replace ("(", "").replace (")","")
            exclude_list[l] = exclude_list[l].split (",")
        item_props.include = include_list
        item_props.exclude = exclude_list
        return item_props`

**Comment by RB-Development** [comment]
I added those missing function in config.py and made some bugfixing regarding to changing function to static_methods in config.py
Should be fine now.

**Comment by RB-Development** [comment]
Hey @aeon0 anymore changes need to be done for merging?

**Comment by Steve6030** [comment]
Using this PR. Working well. Great job.

**Review by mgleed — CHANGES_REQUESTED** [review]
@RB-Development Since aeon reviewed and gave it the go-ahead earlier, I'd be OK merging in his absence. Only thing I'd ask though before merge is this--will you write a better tutorial for how to implement this? Either modify top of pickit.ini further than you have, update your writeup from initial commit with PR, etc. Still a little unsure what I'm supposed to do once I've chosen to use advanced pickit.

**Comment by RB-Development** [comment]
@mgleed added documentary in pickit.ini and changed the inital comment regarding the PR. Hope that matches your expectation :)

**Review by mgleed — APPROVED** [review]

File: src/ui/ui_manager.py
Comment (mgleed) [inline src/ui/ui_manager.py:276]: Some people prefer not to identify items. If they have `id_items=0` shouldn't we set include *and* exclude props to none? Some properties like sockets won't show up on unidentified items iirc.

**Review by mgleed — CHANGES_REQUESTED** [review]
Looks pretty good, and I tested and it works nicely without needing to adjust anything on my end. I think this could potentially be merged and enjoyed by the masses for the next "who knows how long" while we wait for OCR item filtering. I'd better leave it to @aeon0 to decide now because he's back.

**Comment by RB-Development** [comment]
> Looks pretty good, and I tested and it works nicely. I think this could potentially be merged and enjoyed by the masses for the next "who knows how long" while we wait for OCR item filtering. I'd better leave it to @aeon0 to decide now because he's back.

Even with ocr this could be a benefit because you are able to use nested filters. We can throw away the assets part but the Rest should work even with ocr

**Review by mgleed — APPROVED** [review]

**Review by mgleed — APPROVED** [review]

**Comment by RB-Development** [comment]
@mgleed Totally fine to keep gold options to default.

**Pull request merged**
